request.go

```diff
@@ -214,64 +214,77 @@ func (exo *Client) parseResponse(resp *http.Response) (json.RawMessage, error) {
-	if err != nil {
-		return nil, err
-	}
+	totalTime := time.Duration(async.Delay*async.Retries) * time.Second
+	end := time.Now().Add(totalTime)
 
-	// Is it a Job?
-	job := new(JobResultResponse)
-	if err := json.Unmarshal(body, &job); err != nil {
-		return nil, err
-	}
+	channel := make(chan interface{}, 1)
+	go func() {
+		body, err := exo.request(req.name(), req)
+		if err != nil {
+			channel <- err
+			return
+		}
 
-	// Error response
-	errorResponse := new(ErrorResponse)
-	// Successful response
-	resp := req.asyncResponse()
-	if job.JobID == "" || job.JobStatus != Pending {
-		if err := json.Unmarshal(*job.JobResult, resp); err != nil {
-			return job, err
+		// Is it a Job?
+		job := new(JobResultResponse)
+		if err := json.Unmarshal(body, &job); err != nil {
+			channel <- err
+			return
 		}
-		return resp, nil
-	}
 
-	// we've got a pending job
-	result := &QueryAsyncJobResultResponse{
-		JobStatus: job.JobStatus,
-	}
-	for async.Retries > 0 && result.JobStatus == Pending {
-		time.Sleep(time.Duration(async.Delay) * time.Second)
+		// Successful response
+		resp := req.asyncResponse()
+		if job.JobID == "" || job.JobStatus != Pending {
+			if err := json.Unmarshal(*job.JobResult, resp); err != nil {
+				channel <- err
+			}
+			channel <- resp
+			return
+		}
 
-		async.Retries--
+		for async.Retries > 0 {
+			async.Retries--
 
-		req := &QueryAsyncJobResult{JobID: job.JobID}
-		resp, err := exo.Request(req)
-		if err != nil {
-			return nil, err
-		}
-		result = resp.(*QueryAsyncJobResultResponse)
-	}
+			when := end.Add(time.Duration(async.Delay*-async.Retries) * time.Second)
+			time.Sleep(when.Sub(time.Now()))
 
-	if result.JobStatus == Failure {
-		if err := json.Unmarshal(*result.JobResult, &errorResponse); err != nil {
-			return nil, err
+			req := &QueryAsyncJobResult{JobID: job.JobID}
+			resp, err := exo.Request(req)
+			if err != nil {
+				channel <- err
+				return
+			}
+
+			result := resp.(*QueryAsyncJobResultResponse)
+			if result.JobStatus == Success {
+				channel <- result
+			} else if result.JobStatus == Failure {
+				errorResponse := new(ErrorResponse)
+				err := json.Unmarshal(*result.JobResult, &errorResponse)
+				if err != nil {
+					channel <- err
+				}
+				channel <- errorResponse
+				return
+			}
 		}
-		return errorResponse, errorResponse
-	}
 
-	if result.JobStatus == Pending {
-		return result, fmt.Errorf("Maximum number of retries reached")
-	}
+		channel <- fmt.Errorf("Maximum number of retries reached")
+	}()
 
-	if err := json.Unmarshal(*result.JobResult, resp); err != nil {
-		if err := json.Unmarshal(*result.JobResult, errorResponse); err != nil {
-			return nil, err
+	select {
+	case result := <-channel:
+		switch result.(type) {
+		case AsyncJobResult:
+			return result, nil
+		case error:
+			return nil, result.(error)
+		default:
+			panic("Derp")
 		}
-		return errorResponse, errorResponse
+	case <-time.After(totalTime):
+		return nil, fmt.Errorf("Timeout")
```

Comment:
Goroutine would leak. No easy solution, everything needs to be able to be cancelled (usually with the help of a context).